src/resultset/resultset_records.c

```diff
@@ -0,0 +1,171 @@
+    RedisModule_ReplyWithStringBuffer(ctx, "id", 2);
+    RedisModule_ReplyWithLongLong(ctx, n->entity->id);
+
+    // ["labels", [label string]]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "labels", 6);
+    // Print label in nested array for multi-label support
+    RedisModule_ReplyWithArray(ctx, 1);
+    // Retrieve label
+    // TODO Make a more efficient lookup for this string
+    GraphContext *gc = GraphContext_GetFromLTS();
+    int idx = Graph_GetNodeLabel(gc->g, ENTITY_GET_ID(n));
+    if (idx == GRAPH_NO_LABEL) {
+        RedisModule_ReplyWithNull(ctx);
+    } else {
+        const char *label = gc->node_stores[idx]->label;
+        RedisModule_ReplyWithStringBuffer(ctx, label, strlen(label));
+    }
+
+    // [properties, [properties]]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "properties", 10);
+    _enumerateProperties(ctx, n->entity);
+}
+
+void _ResultSet_ReplyWithEdge(RedisModuleCtx *ctx, Edge *e) {
+    // 6 top-level entities in node reply
+    RedisModule_ReplyWithArray(ctx, 6);
+
+    // ["type", "relation"]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "type", 4);
+    RedisModule_ReplyWithStringBuffer(ctx, "relation", 8);
+
+    // ["id", id(int)]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "id", 2);
+    RedisModule_ReplyWithLongLong(ctx, e->entity->id);
+
+    // ["relation_type", type string]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "relation_type", 13);
+    // Retrieve relation type
+    // TODO Make a more efficient lookup for this string
+    GraphContext *gc = GraphContext_GetFromLTS();
+    int idx = Graph_GetEdgeRelation(gc->g, e);
+    const char *label = gc->relation_stores[idx]->label;
+    RedisModule_ReplyWithStringBuffer(ctx, label, strlen(label));
+
+    // ["src_node", srcNodeID(int)]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "src_node", 8);
+    RedisModule_ReplyWithLongLong(ctx, e->srcNodeID);
+
+    // ["dest_node", destNodeID(int)]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "dest_node", 9);
+    RedisModule_ReplyWithLongLong(ctx, e->destNodeID);
+
+    // [properties, [properties]]
+    RedisModule_ReplyWithArray(ctx, 2);
+    RedisModule_ReplyWithStringBuffer(ctx, "properties", 10);
+    _enumerateProperties(ctx, e->entity);
+}
+
+void ResultSet_EmitRecord(RedisModuleCtx *ctx, const Record r, unsigned int numcols) {
+    // Prepare return array sized to the number of RETURN entities
+    RedisModule_ReplyWithArray(ctx, numcols);
+
+    int written;
+    char value[2048];
+    for(int i = 0; i < numcols; i++) {
+        RecordEntryType t = Record_GetType(r, i);
+        switch (t) {
+            case REC_TYPE_SCALAR:
+                /* TODO This logic can be replaced with an _emitSIValue call,
```

Comment:
Haha, sorry - I had meant to bring up this area with you explicitly so that we could discuss it, but failed to!

Right now _emitSIValue also prints the type of the value, but that's easy to tweak.

I haven't written that change in yet because then this PR will modify the behavior of all result sets, not just full-entity results. Specifically, doubles will be converted to strings in the Redis layer, and int and NULL types will be RESP ints and NULLs rather than strings.

That might be fine, but if we dislike that idea, we can do our own string conversions and RedisModuleReply with those.